### PR TITLE
Add HK ADC (ADS1015) functionality on CUBES

### PR DESCRIPTION
**Pull request by sonals301**
Add HK ADC reading for measurement of Voltage, Temperature and Current from CUBES

### CODE REVIEW

File: hk_adc/hk_adc.c [hunk 1]
```diff
@@ -0,0 +1,530 @@
+/*
```
Comment (tstana) [inline hk_adc/hk_adc.c:1]: The usual: GPL header here please.

File: hk_adc/hk_adc.c [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+
+#include "../firmware/drivers/mss_gpio/mss_gpio.h"
+#include "../firmware/drivers/mss_i2c/mss_i2c.h"
+#include "hk_adc.h"
+
+
+#define TX_LENGTH                       1u
+#define RX_LENGTH                       2u
+#define DUMMY_SERIAL_ADDR               0xFF
+#define NUM_SAMPLES_TO_AVG              10u
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Good habit to put the macro assignment value into parantheses, e.g., `#define TX_LENGTH (1u)`.

File: hk_adc/hk_adc.c [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+ *
+ * @param none.
+ * @return HK_ADC_ERR_INIT_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ */
+int hk_adc_init(void)
+{
+    uint8_t err = HK_ADC_ERR_INIT_FAILED;
+
+    // By default, ADC is configured to point towards CITI_TEMP as MUX input
+    uint16_t config = HK_ADC_CONFIG_MUX_SINGLE_2 | HK_ADC_CONFIG_FSR_2 | HK_ADC_CONFIG_MODE_SINGLE_SHOT | HK_ADC_CONFIG_DR_1600SPS | HK_ADC_CONFIG_CMODE_TRAD | HK_ADC_CONFIG_CPOL_ACTIVE_LOW | HK_ADC_CONFIG_CLAT_NON_LATCH | HK_ADC_CONFIG_CQUE_DISABLE;
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Two things about this line:

1. Please code lines within 80 columns, it makes it easier to read without needing to scroll right. (Yes, I know you would then need to scroll down, but that is more natural than scrolling right and left...)
2. Make `config` a `const uint16_t CONFIG` (capitalize it to reflect it is a constant). Seems more logical like this, given it is only used once and within this function.

File: hk_adc/hk_adc.c [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+    uint8_t send_buffer[2] = {config>>8, config};
+
+    // Set the clock for i2c0 instance
+    MSS_I2C_init(&g_mss_i2c0, DUMMY_SERIAL_ADDR, MSS_I2C_PCLK_DIV_256); //10^8/256 = 390,625Hz =~ 390KHz
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Same as above with the 80-column limit. You could put the comment at the end of this line in the already commented line above and fit within 80 cols.

File: hk_adc/hk_adc.c [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+{
+    uint8_t err = HK_ADC_ERR_VOLT_READ_FAILED;
+
+    uint8_t send_buffer[2] = {0};
+
+    err = hk_adc_reg_read(HK_ADC_REG_CONFIG, &read_value);
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        // Let the rest of the configurations remain intact, just change MUX input config and start single-shot conversion
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+            //Wait until OS bit turns 1
+            do{
+                hk_adc_reg_read(HK_ADC_REG_CONFIG, &read_value);
+                os_bit = read_value >> 15;
+            }while(os_bit == HK_ADC_CONFIG_OS_NOT_READY);
+
+            err = hk_adc_reg_read(HK_ADC_REG_CONVERSION, &read_value);
+            if (err == HK_ADC_NO_ERR)
+            {
+                // Right shift the received result by 4 bits. Conversion value are 12 most significant bits
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+/**
+ * @brief HK ADC read Battery Voltage function
+ *
+ * This function returns the raw battery voltage value as read by the ADC. ADC
+ * measures raw Vout value where Vout = V_Batt/8. V_Batt is found by performing
+ * appropriate conversion on the ground software (CITIROC UI tool) side. Raw ADC
+ * value is returned in the unsigned 16-bit parameter passed to the function as
+ * argument.
+ *
+ * @param batt_volt[in]     pointer to unsigned 16-bit variable to store the result in.
```
Comment (tstana) [inline hk_adc/hk_adc.c]: This is actually an `[out]` param.

File: hk_adc/hk_adc.c [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+    return err;
+}
+
+
+
+
+/**
+ * @brief HK ADC read Battery Voltage function
+ *
+ * This function returns the raw battery voltage value as read by the ADC. ADC
```
Comment (tstana) [inline hk_adc/hk_adc.c]: I would rephrase "returns" here, to avoid confusion in the future.

File: hk_adc/hk_adc.c [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+    return err;
+}
+
+
+
+
+/**
+ * @brief HK ADC read Battery Current function
+ *
+ * This function returns the raw battery current value as read by the ADC. ADC
```
Comment (tstana) [inline hk_adc/hk_adc.c]: I would rephrase "returns" here, to avoid confusion in the future.

File: hk_adc/hk_adc.c [hunk 10]
```diff
@@ -0,0 +1,530 @@
...
+ * measures raw Vout value where Vout = I_Batt[Amp] * 2.5. I_Batt is found by performing
+ * appropriate conversion on the ground software (CITIROC UI tool) side. Raw ADC
+ * value is returned in the unsigned 16-bit parameter passed to the function as
+ * argument.
+ *
+ * @param batt_curr[in]     pointer to unsigned 16-bit variable to store the result in.
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Note: `batt_curr[out]`

File: hk_adc/hk_adc.c [hunk 11]
```diff
@@ -0,0 +1,530 @@
...
+    return err;
+}
+
+
+
+
+/**
+ * @brief HK ADC read CITIROC temperature (deg. C) function
+ *
+ * This function returns the raw voltage value (Vtemp) as read by the ADC,
```
Comment (tstana) [inline hk_adc/hk_adc.c]: I would rephrase "returns" here, to avoid confusion in the future.

File: hk_adc/hk_adc.c [hunk 12]
```diff
@@ -0,0 +1,530 @@
...
+ * equivalent to CITIROC ASIC temperature value. Conversion to actual temperature
+ * value is performed by the ground software (CITIROC UI tool).
+ *
+ * @param citi_temp[in]     pointer to unsigned 16-bit variable to store the result in.
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Note: `citi_temp[out]`

File: hk_adc/hk_adc.c [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+{
+    uint8_t err = HK_ADC_ERR_TEMP_READ_FAILED;
+
+    uint8_t send_buffer[2] = {0};
+
+    err = hk_adc_reg_read(HK_ADC_REG_CONFIG, &read_value);
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        // Let the rest of the configurations remain intact, just change MUX input config and start single-shot conversion
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+{
+    uint8_t err = HK_ADC_ERR_CURR_READ_FAILED;
+
+    uint8_t send_buffer[2] = {0};
+
+    err = hk_adc_reg_read(HK_ADC_REG_CONFIG, &read_value);
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        // Let the rest of the configurations remain intact, just change MUX input config and start single-shot conversion
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 15]
```diff
@@ -0,0 +1,530 @@
...
+            //Wait until OS bit turns 1
+            do{
+                hk_adc_reg_read(HK_ADC_REG_CONFIG, &read_value);
+                os_bit = read_value >> 15;
+            }while(os_bit == HK_ADC_CONFIG_OS_NOT_READY);
+
+            err = hk_adc_reg_read(HK_ADC_REG_CONVERSION, &read_value);
+            if (err == HK_ADC_NO_ERR)
+            {
+                // Right shift the received result by 4 bits. Conversion value are 12 most significant bits
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 17]
```diff
@@ -0,0 +1,530 @@
...
+
+//        default:
+            //no other register possible
+    }
+
+    send_buffer[1] = *write_buffer;         //MSB
+    send_buffer[2] = *(write_buffer + 1);   //LSB
+
+    // Write data to slave.
+    MSS_I2C_write(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &send_buffer[0], sizeof(send_buffer)/sizeof(send_buffer[0]), MSS_I2C_RELEASE_BUS);
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 18]
```diff
@@ -0,0 +1,530 @@
...
+        case HK_ADC_REG_HI_THRESH:
+            send_buffer = HK_ADC_REG_PTR_HI_THRESH;
+            break;
+
+//        default:
+            //no other register possible
+    }
+
+    // Write to Address Pointer Register - pointing to 'reg'
+    MSS_I2C_write(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &send_buffer, TX_LENGTH, MSS_I2C_RELEASE_BUS);
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 19]
```diff
@@ -0,0 +1,530 @@
...
+
+    // Wait for completion and record the outcome
+    status = MSS_I2C_wait_complete(&g_mss_i2c0, MSS_I2C_NO_TIMEOUT);
+
+    // Proceed only if it is a success
+    if (status == MSS_I2C_SUCCESS)
+    {
+        // Read data from target slave (we always read 2B)
+        MSS_I2C_read(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &rx_buffer[0], RX_LENGTH, MSS_I2C_RELEASE_BUS);
```
Comment (tstana) [inline hk_adc/hk_adc.c]: 80 cols

File: hk_adc/hk_adc.c [hunk 20]
```diff
@@ -0,0 +1,530 @@
...
+static uint16_t citi_temp_array[NUM_SAMPLES_TO_AVG];
+static uint16_t hk_adc_avg_volt;
+static uint16_t hk_adc_avg_curr;
+static uint16_t hk_adc_avg_citi_temp;
+
+
+/* Local Function prototypes */
+static int hk_adc_reg_write(hk_adc_register_t reg, uint8_t *write_buffer);
+static int hk_adc_reg_read(hk_adc_register_t reg, uint16_t *read_buffer);
+static void hk_adc_update_multiplier_volt(void);
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Function no longer exists, please remove the prototype declaration.

File: hk_adc/hk_adc.c [hunk 21]
```diff
@@ -0,0 +1,530 @@
...
+ *
+ * This local function calculates the average battery voltage from raw ADC voltage
+ * values. Number of samples to be averaged is defined by NUM_SAMPLES_TO_AVG.
+ *
+ * @param None.
+ *
+ * @return Average battery voltage raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_voltage(void)
```
Comment (tstana) [inline hk_adc/hk_adc.c:440]: I see no particular point in having the local `calc` functions. Might as well just move the contents to the `get_` functions, or am I missing something?

File: hk_adc/hk_adc.c [hunk 22]
```diff
@@ -0,0 +1,530 @@
...
+{
+    uint16_t sum = 0;
```
Comment (tstana) [inline hk_adc/hk_adc.c:442]: if  `NUM_SAMPLES_TO_AVG` at some point becomes "16", we run the risk of overflow here... Please see my following comments for suggestions of improvement.

File: hk_adc/hk_adc.c [hunk 23]
```diff
@@ -0,0 +1,530 @@
...
+    uint16_t average = 0;
+    for (int i = 0; i < NUM_SAMPLES_TO_AVG; i ++)
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Suggest we "hard-code" the number of samples to average here, to perhaps 8?

File: hk_adc/hk_adc.c [hunk 24]
```diff
@@ -0,0 +1,530 @@
...
+    {
+        hk_adc_conv_read_volt(&volt_array[i]);
+        sum += volt_array[i];
+        average = (i == NUM_SAMPLES_TO_AVG - 1) ? sum / (uint16_t)NUM_SAMPLES_TO_AVG : 0;
```
Comment (tstana) [inline hk_adc/hk_adc.c]: This seems a bit clunky here... If you set the number of samples to average to **8**, then you could move this statement outside of the for loop and do `average = sum >> 3`, or maybe even `return (sum >> 3)`, avoiding the extra variable?

File: hk_adc/hk_adc.c [hunk 25]
```diff
@@ -0,0 +1,530 @@
...
+ * @return Average battery voltage raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_voltage(void)
+{
+    uint16_t sum = 0;
+    uint16_t average = 0;
+    for (int i = 0; i < NUM_SAMPLES_TO_AVG; i ++)
+    {
+        hk_adc_conv_read_volt(&volt_array[i]);
```
Comment (tstana) [inline hk_adc/hk_adc.c:447]: I would define `volt_array` inside the function, or was there any other reason you defined it as `static`? I see no other use of the variable outside of this function...

File: hk_adc/hk_adc.c [hunk 26]
```diff
@@ -0,0 +1,530 @@
...
+ *
+ * This local function calculates the average battery current from raw ADC voltage
+ * values. Number of samples to be averaged is defined by NUM_SAMPLES_TO_AVG.
+ *
+ * @param None.
+ *
+ * @return Average battery current raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_current(void)
```
Comment (tstana) [inline hk_adc/hk_adc.c:469]: Same comments to this function as to `hk_adc_calc_avg_volt`

File: hk_adc/hk_adc.c [hunk 27]
```diff
@@ -0,0 +1,530 @@
...
+ *
+ * This local function calculates the average CITIROC temperature from raw ADC voltage
+ * values. Number of samples to be averaged is defined by NUM_SAMPLES_TO_AVG.
+ *
+ * @param None.
+ *
+ * @return Average CITIROC temperature raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_citi_temp(void)
```
Comment (tstana) [inline hk_adc/hk_adc.c:498]: Same comments as to `hk_adc_calc_avg_volt()`

File: hk_adc/hk_adc.h [hunk 28]
```diff
@@ -0,0 +1,133 @@
+/*
```
Comment (tstana) [inline hk_adc/hk_adc.h:1]: GPL header. Never mind all the lines > 80 cols, in the interest of saving time, just leave them as they are!

File: msp/msp_exp_handler.c [hunk 29]
```diff
@@ -726,7 +735,7 @@ void msp_exprecv_syscommand(unsigned char opcode)
 void Timer1_IRQHandler(void)
 {
 	citiroc_daq_set_hvps_temp(hvps_get_temp());
-	// citiroc_daq_set_citi_temp(hkadc_read(CITI_TEMP_CHAN));
+    citiroc_daq_set_citi_temp(hk_adc_get_avg_citi_temp());
```
Comment (tstana) [inline msp/msp_exp_handler.c]: Please use a TAB character here.

File: msp/msp_exp_handler.c [hunk 30]
```diff
@@ -413,8 +417,13 @@ void msp_expsend_start(unsigned char opcode, unsigned long *len)
 		send_data_hk[36] = (unsigned char) (count >> 8)  & 0xff;
 		send_data_hk[37] = (unsigned char) (count >> 0)  & 0xff;
 
+        /* ADC HK */
+        sprintf(((char*)send_data_hk)+38, "%04u", hk_adc_get_avg_volt());
+        sprintf(((char*)send_data_hk)+42, "%04u", hk_adc_get_avg_curr());
+        sprintf(((char*)send_data_hk)+46, "%04u", hk_adc_get_avg_citi_temp());
```
Comment (tstana) [inline msp/msp_exp_handler.c]: Please use TAB characters here...

**Review by tstana — CHANGES_REQUESTED** [review]
Hi Sonal, I went through your code and made some comments. Hope they won't demoralize you too much, most of them are asthaetic and pretty straightforward to go through. Great job on the functions overall, amazing that they are working!

File: hk_adc/hk_adc.c (hunk 1 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:1]: Done

File: hk_adc/hk_adc.c (hunk 2 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 3 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Good points! I made the change

File: hk_adc/hk_adc.c (hunk 4 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Haha yeah, sorry! I don't remember why I even put it there in the first place. My bad! 
Made the change.

File: hk_adc/hk_adc.c (hunk 5 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 6 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 7 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: True! Made the change.

File: hk_adc/hk_adc.c (hunk 8 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Made the change. Hope it looks better now.

File: hk_adc/hk_adc.c (hunk 9 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Same as above

File: hk_adc/hk_adc.c (hunk 10 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 11 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Same as above

File: hk_adc/hk_adc.c (hunk 12 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 13 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 14 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 15 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 16 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 17 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 18 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 19 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 20 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Sorry, missed out on this one previously. Removed it now.

File: hk_adc/hk_adc.c (hunk 21 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:440]: Nothing big, just that the `get_` fns act as exported interface fns to get the calculated avg values. I have removed this and exposed the `calc_` fns instead. Hope it is okay now.

File: hk_adc/hk_adc.c (hunk 22 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:433]: Done

File: hk_adc/hk_adc.c (hunk 24 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Okay, sounds good to me. Made the change

File: hk_adc/hk_adc.c (hunk 25 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:447]: True, I made it local now

File: hk_adc/hk_adc.c (hunk 26 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:469]: Done

File: hk_adc/hk_adc.c (hunk 27 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:498]: Done

File: hk_adc/hk_adc.h (hunk 28 above)
Comment (sonals301) [reply hk_adc/hk_adc.h:1]: I made the changes :) It did not take much time

File: msp/msp_exp_handler.c (hunk 29 above)
Comment (sonals301) [reply msp/msp_exp_handler.c]: I could not see this in my Eclipse IDE. I changed the settings and now I can see tabs and space characters. Made the change.

File: msp/msp_exp_handler.c (hunk 30 above)
Comment (sonals301) [reply msp/msp_exp_handler.c]: Done

File: hk_adc/hk_adc.c (hunk 23 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Done

File: hk_adc/hk_adc.c (hunk 22 above)
Comment (sonals301) [reply hk_adc/hk_adc.c:442]: Made the change as per other comments.

**Comment by sonals301** [comment]
Hi Thedi,
I addressed all your comments. Please pardon me for so many silly mistakes I made. I hope it is all better now. Let me know if there is anything else.

File: hk_adc/hk_adc.c [hunk 31]
```diff
@@ -0,0 +1,504 @@
...
+ *
+ * @param reg[in]              Variable of enumeration type #hk_adc_register_t
+ *                             (defined in hk_adc.h) indicating which register
+ *                             is to be read.
+ * @param read_buffer[in]      pointer to a uint16_t variable to store the result
+ *                             in.
+ * @return HK_ADC_ERR_READ_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ *
+ */
+int hk_adc_reg_read(hk_adc_register_t reg, uint16_t *read_buffer)
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Missing a `static` here (for clarity, it would probably not be an issue globally).

File: hk_adc/hk_adc.c [hunk 32]
```diff
@@ -0,0 +1,504 @@
...
+ *
+ * @param reg[in]              Variable of enumeration type #hk_adc_register_t
+ *                             (defined in hk_adc.h) indicating which register
+ *                             is to be written.
+ * @param write_buffer[in]     pointer to an unsigned byte array containing data
+ *                             to be written.
+ * @return HK_ADC_ERR_WRITE_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ *
+ */
+int hk_adc_reg_write(hk_adc_register_t reg, uint8_t *write_buffer)
```
Comment (tstana) [inline hk_adc/hk_adc.c]: Missing a `static` here (for clarity, it would probably not be an issue globally).

File: hk_adc/hk_adc.h (hunk 28 above)
Comment (tstana) [reply hk_adc/hk_adc.h:1]: Thank you, much appreciated -- and it looks much better now! 🙂

File: hk_adc/hk_adc.c [hunk 33]
```diff
@@ -0,0 +1,504 @@
...
+
+
+/* Local Variables */
+static mss_i2c_status_t status;
+static uint16_t read_value;
+static uint8_t os_bit = 0;
+static uint8_t  rx_buffer[RX_LENGTH];
+static uint16_t hk_adc_avg_volt;
+static uint16_t hk_adc_avg_curr;
+static uint16_t hk_adc_avg_citi_temp;
```
Comment (tstana) [inline hk_adc/hk_adc.c]: These var's are all only used within one function, so they can be made local to these functions. It would also feel safer to me if the variables above them would also be made local to the functions they are used in, there is less risk that if one function gets changed at a later point, it breaks another's assignment to the variable. But if you have no time to address the variables above, then just open up an issue on it and let it be.

**Review by tstana — CHANGES_REQUESTED** [review]
No worries, they weren't silly mistakes at all! I found some more light issues to be addressed, hopefully they won't take too long.

File: hk_adc/hk_adc.c (hunk 32 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Added now.

File: hk_adc/hk_adc.c (hunk 31 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: Yeah, sorry. Added now.

File: hk_adc/hk_adc.c (hunk 33 above)
Comment (sonals301) [reply hk_adc/hk_adc.c]: I made these variables local. But I think the `_avg_` variables should be left global because they are returned outside this file when these fns are called in `msp_exphandler.c`.

Comment (tstana) [reply hk_adc/hk_adc.c]: No, they are not needed. When you return from the `_calc_avg_` functions, they are effectively deallocated and the return value stored in a processor register. When the functions are called from `msp_exp_handler.c`, the call allocates the return value from a processor register to the `sprintf` parameter.

But no need to worry about it too much, it is just a minor nice-to-have detail... I'll just open an issue on this and mark this conversation as resolved.
